### PR TITLE
[SPARK-50409][SQL] Fix set statement to ignore `;` at the end of `SET;`, `SET -v;` and `SET key;`

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
This PR aims to fix the problem of trailing `;` in set statement.

### Why are the changes needed?
We missed to add `;*` in the regex rules in SparkSqlParser.scala.
Several queries were failing:

```
---------------------------------------
set;
---------------------------------------
[INVALID_SET_SYNTAX] Expected format is 'SET', 'SET key', or 'SET key=value'. If you want to include special characters in key, or include semicolon in value, please use backquotes, e.g., SET `key`=`value`. SQLSTATE: 42000
== SQL (line 1, position 1) ==
set;
^^^
---------------------------------------
set -v;
---------------------------------------
[INVALID_SET_SYNTAX] Expected format is 'SET', 'SET key', or 'SET key=value'. If you want to include special characters in key, or include semicolon in value, please use backquotes, e.g., SET `key`=`value`. SQLSTATE: 42000
== SQL (line 1, position 1) ==
set -v;
^^^^^^^
---------------------------------------
set spark.sql.ansi.enabled;
---------------------------------------
[INVALID_SET_SYNTAX] Expected format is 'SET', 'SET key', or 'SET key=value'. If you want to include special characters in key, or include semicolon in value, please use backquotes, e.g., SET `key`=`value`. SQLSTATE: 42000
== SQL (line 1, position 1) ==
set spark.sql.ansi.enabled;
^^^^^^^^^^^^^^^^^^^^^^^^^^
```

All of these would get the `;` at the end and fail due to impossible match.

### Does this PR introduce _any_ user-facing change?
Yes. This is a bug that we had for a long time.

### How was this patch tested?
Added test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by mihailom-db** [comment]
@MaxGekk @cloud-fan please review this. This is an important fix for us, as we have queries failing. One example of the failed query is `set;`

**Comment by cloud-fan** [comment]
How do they fail?

**Comment by mihailom-db** [comment]
Updated with text from failure.

**Review by cloud-fan — APPROVED** [review]

**Review by srielau — APPROVED** [review]

**Review by MaxGekk — COMMENTED** [review]
@mihailom-db Could you fix the failed tests, seems they are related to your changes:
```
[info] - INVALID_PROPERTY_VALUE: invalid property value for set quoted configuration *** FAILED *** (23 milliseconds)
[info]   Map("value" -> ""1;2"", "key" -> ""key"") did not equal Map("value" -> ""1;2;;"", "key" -> ""key"") (SparkFunSuite.scala:362)
[info]   Analysis:
[info]   JavaCollectionWrappers$JMapWrapper("value": ""1;2"" -> ""1;2;;"")
[info]   org.scalatest.exceptions.TestFailedException:
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 1]
```diff
@@ -2175,4 +2175,9 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
       sql(s"CREATE TABLE t (c STRING COLLATE system.builtin.UTF8_LCASE)")
     }
   }
+
+  test("dsadasda") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:2179]: nice name for the test. How is it related to collation?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -106,14 +106,14 @@ class SparkSqlAstBuilder extends AstBuilder {
         SetCommand(Some(keyStr -> None))
       }
     } else {
-      remainder(ctx.SET.getSymbol).trim match {
+      remainder(ctx.SET.getSymbol).trim.replaceAll(";+$", "") match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:109]: is it the same as `.trimRight(";")`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala [hunk 3]
```diff
@@ -93,6 +93,15 @@ class SparkSqlParserSuite extends AnalysisTest with SharedSparkSession {
       parameters = Map.empty)
   }
 
+  test("SET with semi-colons") {
+    assertEqual(s"SET;", SetCommand(None))
+    assertEqual(s"SET    ;", SetCommand(None))
+    assertEqual(s"SET -v;", SetCommand(Some("-v" -> None)))
+    assertEqual(s"SET -v    ;", SetCommand(Some("-v" -> None)))
+    assertEqual(s"SET spark.sql.ansi.enabled;", SetCommand(Some("spark.sql.ansi.enabled" -> None)))
+    assertEqual(s"SET spark.sql.ansi.enabled ;", SetCommand(Some("spark.sql.ansi.enabled" -> None)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:102]: What were the behavior for these commands before this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 2 above)
Comment (mihailom-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:109]: Well, `.trimRight(";")` can only be done on `UTF8String`, but here we have java `String`. In java we do not have formal trimRight or trimLeft, but only trim for spaces. This is a regex replace to exchange all occurrences of `;` at the end ($) of the String.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala (hunk 3 above)
Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:102]: All failed with PARSE_SYNTAX_ERROR.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 1 above)
Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: This was accidental push, reverted.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**